File: src/main/java/dev/esophose/discordbot/command/commands/moderation/KickCommand.java
package dev.esophose.discordbot.command.commands.moderation;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import dev.esophose.discordbot.manager.CommandManager;
import discord4j.core.object.entity.Member;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class KickCommand extends DiscordCommand {

    public KickCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, Member member, Optional<String> reason) {
        CommandManager commandManager = this.bot.getManager(CommandManager.class);

        message.getGuild()
                .flatMap(x -> reason.map(s -> x.kick(member.getId(), s)).orElseGet(() -> x.kick(member.getId())))
                .doOnError(error -> commandManager.sendResponse(message.getChannel(), "Failed to kick user", "An error occurred trying to kick that user: " + error.getMessage()).subscribe())
                .doOnSuccess(v -> commandManager.sendResponse(message.getChannel(), "User kicked", member.getMention() + " has been kicked.").subscribe())
                .subscribe();
    }

    @Override
    public String getName() {
        return "kick";
    }

    @Override
    public List<String> getAliases() {
        return Collections.emptyList();
    }

    @Override
    public String getDescription() {
        return "Kicks a member";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES, Permission.KICK_MEMBERS);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.KICK_MEMBERS;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/moderation/KickCommand.kt
package dev.esophose.discordbot.command.commands.moderation

import dev.esophose.discordbot.Sparky
import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import dev.esophose.discordbot.manager.CommandManager
import discord4j.core.`object`.entity.Member
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet
import java.util.Optional

class KickCommand : DiscordCommand() {

    override val name: String
        get() = "kick"

    override val aliases: List<String>
        get() = emptyList()

    override val description: String
        get() = "Kicks a member"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES, Permission.KICK_MEMBERS)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.KICK_MEMBERS

    fun execute(message: DiscordCommandMessage, member: Member, reason: Optional<String>) {
        val commandManager = Sparky.getManager(CommandManager::class)

        if (member.id == Sparky.self.id) {
            commandManager.sendResponse(message.channel, "Why would you do that?", "Why would you try to kick me? What have I ever done to you?").subscribe()
            return
        }

        message.guild
                .flatMap { x -> reason.map { s -> x.kick(member.id, s) }.orElseGet { x.kick(member.id) } }
                .doOnError { error -> commandManager.sendResponse(message.channel, "Failed to kick user", "An error occurred trying to kick that user: " + error.message).subscribe() }
                .doOnSuccess { commandManager.sendResponse(message.channel, "User kicked", member.mention + " has been kicked.").subscribe() }
                .subscribe()
    }

}
